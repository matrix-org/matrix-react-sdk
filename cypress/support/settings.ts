/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/// <reference types="cypress" />

import Chainable = Cypress.Chainable;
import Loggable = Cypress.Loggable;
import Timeoutable = Cypress.Timeoutable;
import Withinable = Cypress.Withinable;
import Shadow = Cypress.Shadow;
import type { SettingLevel } from "../../src/settings/SettingLevel";
import type SettingsStore from "../../src/settings/SettingsStore";

export enum Filter {
    People = "people",
    PublicRooms = "public_rooms",
}

declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace Cypress {
        interface Chainable {
            /**
             * Returns the SettingsStore
             */
            getSettingsStore(): Chainable<typeof SettingsStore | undefined>; // XXX: Importing SettingsStore causes a bunch of type lint errors
            /**
             * Open the top left user menu, returning a handle to the resulting context menu.
             */
            openUserMenu(): Chainable<JQuery<HTMLElement>>;

            /**
             * Open user settings (via user menu), returning a handle to the resulting dialog.
             * @param tab the name of the tab to switch to after opening, optional.
             */
            openUserSettings(tab?: string): Chainable<JQuery<HTMLElement>>;

            /**
             * Open room creation dialog.
             */
            openCreateRoomDialog(): Chainable<JQuery<HTMLElement>>;

            /**
             * Open room settings (via room header menu), returning a handle to the resulting dialog.
             * @param tab the name of the tab to switch to after opening, optional.
             */
            openRoomSettings(tab?: string): Chainable<JQuery<HTMLElement>>;

            /**
             * Switch settings tab to the one by the given name, ideally call this in the context of the dialog.
             * @param tab the name of the tab to switch to.
             */
            switchTab(tab: string): Chainable<JQuery<HTMLElement>>;

            /**
             * Close dialog, ideally call this in the context of the dialog.
             */
            closeDialog(): Chainable<JQuery<HTMLElement>>;

            /**
             * Join the given beta, the `Labs` tab must already be opened,
             * ideally call this in the context of the dialog.
             * @param name the name of the beta to join.
             */
            joinBeta(name: string): Chainable<JQuery<HTMLElement>>;

            /**
             * Leave the given beta, the `Labs` tab must already be opened,
             * ideally call this in the context of the dialog.
             * @param name the name of the beta to leave.
             */
            leaveBeta(name: string): Chainable<JQuery<HTMLElement>>;

            /**
             * Sets the value for a setting. The room ID is optional if the
             * setting is not being set for a particular room, otherwise it
             * should be supplied. The value may be null to indicate that the
             * level should no longer have an override.
             * @param {string} settingName The name of the setting to change.
             * @param {String} roomId The room ID to change the value in, may be
             * null.
             * @param {SettingLevel} level The level to change the value at.
             * @param {*} value The new value of the setting, may be null.
             * @return {Promise} Resolves when the setting has been changed.
             */
            setSettingValue(settingName: string, roomId: string, level: SettingLevel, value: any): Chainable<void>;

            /**
             * Gets the value of a setting. The room ID is optional if the
             * setting is not to be applied to any particular room, otherwise it
             * should be supplied.
             * @param {string} settingName The name of the setting to read the
             * value of.
             * @param {String} roomId The room ID to read the setting value in,
             * may be null.
             * @param {boolean} excludeDefault True to disable using the default
             * value.
             * @return {*} The value, or null if not found
             */
            getSettingValue<T>(settingName: string, roomId?: string, excludeDefault?: boolean): Chainable<T>;

            /**
             * Opens the spotlight dialog
             */
            openSpotlightDialog(
                options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
            ): Chainable<JQuery<HTMLElement>>;
            spotlightDialog(
                options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
            ): Chainable<JQuery<HTMLElement>>;
            spotlightFilter(
                filter: Filter | null,
                options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
            ): Chainable<JQuery<HTMLElement>>;
            spotlightSearch(
                options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
            ): Chainable<JQuery<HTMLElement>>;
            spotlightResults(
                options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
            ): Chainable<JQuery<HTMLElement>>;
        }
    }
}

Cypress.Commands.add("getSettingsStore", (): Chainable<typeof SettingsStore> => {
    return cy.window({ log: false }).then((win) => win.mxSettingsStore);
});

Cypress.Commands.add(
    "setSettingValue",
    (name: string, roomId: string, level: SettingLevel, value: any): Chainable<void> => {
        return cy.getSettingsStore().then((store: typeof SettingsStore) => {
            return cy.wrap(store.setValue(name, roomId, level, value));
        });
    },
);

// eslint-disable-next-line max-len
Cypress.Commands.add(
    "getSettingValue",
    <T = any>(name: string, roomId?: string, excludeDefault?: boolean): Chainable<T> => {
        return cy.getSettingsStore().then((store: typeof SettingsStore) => {
            return store.getValue(name, roomId, excludeDefault);
        });
    },
);

Cypress.Commands.add("openUserMenu", (): Chainable<JQuery<HTMLElement>> => {
    cy.findByRole("button", { name: "User menu" }).click();
    return cy.get(".mx_ContextualMenu");
});

Cypress.Commands.add("openUserSettings", (tab?: string): Chainable<JQuery<HTMLElement>> => {
    cy.openUserMenu().within(() => {
        cy.findByRole("menuitem", { name: "All settings" }).click();
    });
    return cy.get(".mx_UserSettingsDialog").within(() => {
        if (tab) {
            cy.switchTab(tab);
        }
    });
});

Cypress.Commands.add("openCreateRoomDialog", (): Chainable<JQuery<HTMLElement>> => {
    cy.findByRole("button", { name: "Add room" }).click();
    cy.findByRole("menuitem", { name: "New room" }).click();
    return cy.get(".mx_CreateRoomDialog");
});

Cypress.Commands.add("openRoomSettings", (tab?: string): Chainable<JQuery<HTMLElement>> => {
    cy.findByRole("button", { name: "Room options" }).click();
    cy.get(".mx_RoomTile_contextMenu").within(() => {
        cy.findByRole("menuitem", { name: "Settings" }).click();
    });
    return cy.get(".mx_RoomSettingsDialog").within(() => {
        if (tab) {
            cy.switchTab(tab);
        }
    });
});

Cypress.Commands.add("switchTab", (tab: string): Chainable<JQuery<HTMLElement>> => {
    return cy.get(".mx_TabbedView_tabLabels").within(() => {
        cy.contains(".mx_TabbedView_tabLabel", tab).click();
    });
});

Cypress.Commands.add("closeDialog", (): Chainable<JQuery<HTMLElement>> => {
    return cy.findByRole("button", { name: "Close dialog" }).click();
});

Cypress.Commands.add("joinBeta", (name: string): Chainable<JQuery<HTMLElement>> => {
    return cy
        .contains(".mx_BetaCard_title", name)
        .closest(".mx_BetaCard")
        .within(() => {
            return cy.get(".mx_BetaCard_buttons").findByRole("button", { name: "Join the beta" }).click();
        });
});

Cypress.Commands.add("leaveBeta", (name: string): Chainable<JQuery<HTMLElement>> => {
    return cy
        .contains(".mx_BetaCard_title", name)
        .closest(".mx_BetaCard")
        .within(() => {
            return cy.get(".mx_BetaCard_buttons").findByRole("button", { name: "Leave the beta" }).click();
        });
});

Cypress.Commands.add(
    "openSpotlightDialog",
    (options?: Partial<Loggable & Timeoutable & Withinable & Shadow>): Chainable<JQuery<HTMLElement>> => {
        cy.get(".mx_RoomSearch_spotlightTrigger", options).click({ force: true });
        return cy.spotlightDialog(options);
    },
);

Cypress.Commands.add(
    "spotlightDialog",
    (options?: Partial<Loggable & Timeoutable & Withinable & Shadow>): Chainable<JQuery<HTMLElement>> => {
        return cy.get('[role=dialog][aria-label="Search Dialog"]', options);
    },
);

Cypress.Commands.add(
    "spotlightFilter",
    (
        filter: Filter | null,
        options?: Partial<Loggable & Timeoutable & Withinable & Shadow>,
    ): Chainable<JQuery<HTMLElement>> => {
        let selector: string;
        switch (filter) {
            case Filter.People:
                selector = "#mx_SpotlightDialog_button_startChat";
                break;
            case Filter.PublicRooms:
                selector = "#mx_SpotlightDialog_button_explorePublicRooms";
                break;
            default:
                selector = ".mx_SpotlightDialog_filter";
                break;
        }
        return cy.get(selector, options).click();
    },
);

Cypress.Commands.add(
    "spotlightSearch",
    (options?: Partial<Loggable & Timeoutable & Withinable & Shadow>): Chainable<JQuery<HTMLElement>> => {
        return cy.get(".mx_SpotlightDialog_searchBox", options).findByRole("textbox", { name: "Search" });
    },
);

Cypress.Commands.add(
    "spotlightResults",
    (options?: Partial<Loggable & Timeoutable & Withinable & Shadow>): Chainable<JQuery<HTMLElement>> => {
        return cy.get(".mx_SpotlightDialog_section.mx_SpotlightDialog_results .mx_SpotlightDialog_option", options);
    },
);

// Needed to make this file a module
export {};
