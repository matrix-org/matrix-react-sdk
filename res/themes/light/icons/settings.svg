<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M452.515 237l31.843-18.382c9.426-5.441 13.996-16.542 11.177-27.054-11.404-42.531-33.842-80.547-64.058-110.797-7.68-7.688-19.575-9.246-28.985-3.811l-31.785 18.358a196.276 196.276 0 0 0-32.899-19.02V39.541a24.016 24.016 0 0 0-17.842-23.206c-41.761-11.107-86.117-11.121-127.93-.001-10.519 2.798-17.844 12.321-17.844 23.206v36.753a196.276 196.276 0 0 0-32.899 19.02l-31.785-18.358c-9.41-5.435-21.305-3.877-28.985 3.811-30.216 30.25-52.654 68.265-64.058 110.797-2.819 10.512 1.751 21.613 11.177 27.054L59.485 237a197.715 197.715 0 0 0 0 37.999l-31.843 18.382c-9.426 5.441-13.996 16.542-11.177 27.054 11.404 42.531 33.842 80.547 64.058 110.797 7.68 7.688 19.575 9.246 28.985 3.811l31.785-18.358a196.202 196.202 0 0 0 32.899 19.019v36.753a24.016 24.016 0 0 0 17.842 23.206c41.761 11.107 86.117 11.122 127.93.001 10.519-2.798 17.844-12.321 17.844-23.206v-36.753a196.34 196.34 0 0 0 32.899-19.019l31.785 18.358c9.41 5.435 21.305 3.877 28.985-3.811 30.216-30.25 52.654-68.266 64.058-110.797 2.819-10.512-1.751-21.613-11.177-27.054L452.515 275c1.22-12.65 1.22-25.35 0-38zm-52.679 63.019l43.819 25.289a200.138 200.138 0 0 1-33.849 58.528l-43.829-25.309c-31.984 27.397-36.659 30.077-76.168 44.029v50.599a200.917 200.917 0 0 1-67.618 0v-50.599c-39.504-13.95-44.196-16.642-76.168-44.029l-43.829 25.309a200.15 200.15 0 0 1-33.849-58.528l43.819-25.289c-7.63-41.299-7.634-46.719 0-88.038l-43.819-25.289c7.85-21.229 19.31-41.049 33.849-58.529l43.829 25.309c31.984-27.397 36.66-30.078 76.168-44.029V58.845a200.917 200.917 0 0 1 67.618 0v50.599c39.504 13.95 44.196 16.642 76.168 44.029l43.829-25.309a200.143 200.143 0 0 1 33.849 58.529l-43.819 25.289c7.631 41.3 7.634 46.718 0 88.037zM256 160c-52.935 0-96 43.065-96 96s43.065 96 96 96 96-43.065 96-96-43.065-96-96-96zm0 144c-26.468 0-48-21.532-48-48 0-26.467 21.532-48 48-48s48 21.533 48 48c0 26.468-21.532 48-48 48z"/></svg>