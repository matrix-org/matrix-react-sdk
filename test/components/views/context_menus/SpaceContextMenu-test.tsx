/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
// eslint-disable-next-line deprecate/import
import { mount } from 'enzyme';
import { Room } from 'matrix-js-sdk/src/matrix';
import { mocked } from 'jest-mock';
import { act } from 'react-dom/test-utils';
import 'focus-visible'; // to fix context menus

import SpaceContextMenu from '../../../../src/components/views/context_menus/SpaceContextMenu';
import MatrixClientContext from '../../../../src/contexts/MatrixClientContext';
import { findByTestId } from '../../../test-utils';
import {
    shouldShowSpaceSettings,
    showCreateNewRoom,
    showCreateNewSubspace,
    showSpaceInvite,
    showSpaceSettings,
} from '../../../../src/utils/space';
import { leaveSpace } from "../../../../src/utils/leave-behaviour";
import { shouldShowComponent } from '../../../../src/customisations/helpers/UIComponents';
import { UIComponent } from '../../../../src/settings/UIFeature';

jest.mock('../../../../src/customisations/helpers/UIComponents', () => ({
    shouldShowComponent: jest.fn(),
}));

jest.mock('../../../../src/utils/space', () => ({
    shouldShowSpaceSettings: jest.fn(),
    showCreateNewRoom: jest.fn(),
    showCreateNewSubspace: jest.fn(),
    showSpaceInvite: jest.fn(),
    showSpacePreferences: jest.fn(),
    showSpaceSettings: jest.fn(),
}));

jest.mock('../../../../src/utils/leave-behaviour', () => ({
    leaveSpace: jest.fn(),
}));

describe('<SpaceContextMenu />', () => {
    const userId = '@test:server';
    const mockClient = {
        getUserId: jest.fn().mockReturnValue(userId),
    };
    const makeMockSpace = (props = {}) => ({
        name: 'test space',
        getJoinRule: jest.fn(),
        canInvite: jest.fn(),
        currentState: {
            maySendStateEvent: jest.fn(),
        },
        client: mockClient,
        getMyMembership: jest.fn(),
        ...props,
    }) as unknown as Room;
    const defaultProps = {
        space: makeMockSpace(),
        onFinished: jest.fn(),
    };
    const getComponent = (props = {}) =>
        mount(<SpaceContextMenu {...defaultProps} {...props} />,
            {
                wrappingComponent: MatrixClientContext.Provider,
                wrappingComponentProps: {
                    value: mockClient,
                },
            });

    beforeEach(() => {
        jest.resetAllMocks();
        mockClient.getUserId.mockReturnValue(userId);
    });

    it('renders menu correctly', () => {
        const component = getComponent();
        expect(component).toMatchSnapshot();
    });

    it('renders invite option when space is public', () => {
        const space = makeMockSpace({
            getJoinRule: jest.fn().mockReturnValue('public'),
        });
        const component = getComponent({ space });
        expect(findByTestId(component, 'invite-option').length).toBeTruthy();
    });
    it('renders invite option when user is has invite rights for space', () => {
        const space = makeMockSpace({
            canInvite: jest.fn().mockReturnValue(true),
        });
        const component = getComponent({ space });
        expect(space.canInvite).toHaveBeenCalledWith(userId);
        expect(findByTestId(component, 'invite-option').length).toBeTruthy();
    });
    it('opens invite dialog when invite option is clicked', () => {
        const space = makeMockSpace({
            getJoinRule: jest.fn().mockReturnValue('public'),
        });
        const onFinished = jest.fn();
        const component = getComponent({ space, onFinished });

        act(() => {
            findByTestId(component, 'invite-option').at(0).simulate('click');
        });

        expect(showSpaceInvite).toHaveBeenCalledWith(space);
        expect(onFinished).toHaveBeenCalled();
    });
    it('renders space settings option when user has rights', () => {
        mocked(shouldShowSpaceSettings).mockReturnValue(true);
        const component = getComponent();
        expect(shouldShowSpaceSettings).toHaveBeenCalledWith(defaultProps.space);
        expect(findByTestId(component, 'settings-option').length).toBeTruthy();
    });
    it('opens space settings when space settings option is clicked', () => {
        mocked(shouldShowSpaceSettings).mockReturnValue(true);
        const onFinished = jest.fn();
        const component = getComponent({ onFinished });

        act(() => {
            findByTestId(component, 'settings-option').at(0).simulate('click');
        });

        expect(showSpaceSettings).toHaveBeenCalledWith(defaultProps.space);
        expect(onFinished).toHaveBeenCalled();
    });
    it('renders leave option when user does not have rights to see space settings', () => {
        const component = getComponent();
        expect(findByTestId(component, 'leave-option').length).toBeTruthy();
    });
    it('leaves space when leave option is clicked', () => {
        const onFinished = jest.fn();
        const component = getComponent({ onFinished });
        act(() => {
            findByTestId(component, 'leave-option').at(0).simulate('click');
        });
        expect(leaveSpace).toHaveBeenCalledWith(defaultProps.space);
        expect(onFinished).toHaveBeenCalled();
    });
    describe('add children section', () => {
        const space = makeMockSpace();
        beforeEach(() => {
            // set space to allow adding children to space
            mocked(space.currentState.maySendStateEvent).mockReturnValue(true);
            mocked(shouldShowComponent).mockReturnValue(true);
        });
        it('does not render section when user does not have permission to add children', () => {
            mocked(space.currentState.maySendStateEvent).mockReturnValue(false);
            const component = getComponent({ space });

            expect(findByTestId(component, 'add-to-space-header').length).toBeFalsy();
            expect(findByTestId(component, 'new-room-option').length).toBeFalsy();
            expect(findByTestId(component, 'new-subspace-option').length).toBeFalsy();
        });
        it('does not render section when UIComponent customisations disable room and space creation', () => {
            mocked(shouldShowComponent).mockReturnValue(false);
            const component = getComponent({ space });

            expect(shouldShowComponent).toHaveBeenCalledWith(UIComponent.CreateRooms);
            expect(shouldShowComponent).toHaveBeenCalledWith(UIComponent.CreateSpaces);

            expect(findByTestId(component, 'add-to-space-header').length).toBeFalsy();
            expect(findByTestId(component, 'new-room-option').length).toBeFalsy();
            expect(findByTestId(component, 'new-subspace-option').length).toBeFalsy();
        });

        it('renders section with add room button when UIComponent customisation allows CreateRoom', () => {
            // only allow CreateRoom
            mocked(shouldShowComponent).mockImplementation(feature => feature === UIComponent.CreateRooms);
            const component = getComponent({ space });

            expect(findByTestId(component, 'add-to-space-header').length).toBeTruthy();
            expect(findByTestId(component, 'new-room-option').length).toBeTruthy();
            expect(findByTestId(component, 'new-subspace-option').length).toBeFalsy();
        });

        it('renders section with add space button when UIComponent customisation allows CreateSpace', () => {
            // only allow CreateSpaces
            mocked(shouldShowComponent).mockImplementation(feature => feature === UIComponent.CreateSpaces);
            const component = getComponent({ space });

            expect(findByTestId(component, 'add-to-space-header').length).toBeTruthy();
            expect(findByTestId(component, 'new-room-option').length).toBeFalsy();
            expect(findByTestId(component, 'new-subspace-option').length).toBeTruthy();
        });

        it('opens create room dialog on add room button click', () => {
            const onFinished = jest.fn();
            const component = getComponent({ space, onFinished });

            act(() => {
                findByTestId(component, 'new-room-option').at(0).simulate('click');
            });
            expect(showCreateNewRoom).toHaveBeenCalledWith(space);
            expect(onFinished).toHaveBeenCalled();
        });
        it('opens create space dialog on add space button click', () => {
            const onFinished = jest.fn();
            const component = getComponent({ space, onFinished });

            act(() => {
                findByTestId(component, 'new-subspace-option').at(0).simulate('click');
            });
            expect(showCreateNewSubspace).toHaveBeenCalledWith(space);
            expect(onFinished).toHaveBeenCalled();
        });
    });
});
