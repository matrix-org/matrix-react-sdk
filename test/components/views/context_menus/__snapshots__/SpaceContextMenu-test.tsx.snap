// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpaceContextMenu /> renders menu correctly 1`] = `
<SpaceContextMenu
  onFinished={[MockFunction]}
  space={
    Object {
      "canInvite": [MockFunction] {
        "calls": Array [
          Array [
            "@test:server",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "client": Object {
        "getUserId": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": "@test:server",
            },
          ],
        },
      },
      "currentState": Object {
        "maySendStateEvent": [MockFunction] {
          "calls": Array [
            Array [
              "m.space.child",
              "@test:server",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      "getJoinRule": [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "getMyMembership": [MockFunction],
      "name": "test space",
    }
  }
>
  <IconizedContextMenu
    className="mx_SpacePanel_contextMenu"
    compact={true}
    onFinished={[MockFunction]}
  >
    <ContextMenu
      chevronFace="none"
      hasBackground={true}
      managed={true}
      onFinished={[MockFunction]}
    >
      <Portal
        containerInfo={
          <div
            id="mx_ContextualMenu_Container"
          >
            <div
              class="mx_ContextualMenu_wrapper"
            >
              <div
                class="mx_ContextualMenu_background"
              />
              <div
                class="mx_ContextualMenu"
                role="menu"
              >
                <div
                  class="mx_IconizedContextMenu mx_SpacePanel_contextMenu mx_IconizedContextMenu_compact"
                >
                  <div
                    class="mx_SpacePanel_contextMenu_header"
                  >
                    test space
                  </div>
                  <div
                    class="mx_IconizedContextMenu_optionList"
                  >
                    <div
                      aria-label="Space home"
                      class="mx_AccessibleButton mx_IconizedContextMenu_item focus-visible"
                      data-focus-visible-added=""
                      role="menuitem"
                      tabindex="0"
                    >
                      <span
                        class="mx_IconizedContextMenu_icon mx_SpacePanel_iconHome"
                      />
                      <span
                        class="mx_IconizedContextMenu_label"
                      >
                        Space home
                      </span>
                    </div>
                    <div
                      aria-label="Explore rooms"
                      class="mx_AccessibleButton mx_IconizedContextMenu_item"
                      role="menuitem"
                      tabindex="-1"
                    >
                      <span
                        class="mx_IconizedContextMenu_icon mx_SpacePanel_iconExplore"
                      />
                      <span
                        class="mx_IconizedContextMenu_label"
                      >
                        Explore rooms
                      </span>
                    </div>
                    <div
                      aria-label="Preferences"
                      class="mx_AccessibleButton mx_IconizedContextMenu_item"
                      role="menuitem"
                      tabindex="-1"
                    >
                      <span
                        class="mx_IconizedContextMenu_icon mx_SpacePanel_iconPreferences"
                      />
                      <span
                        class="mx_IconizedContextMenu_label"
                      >
                        Preferences
                      </span>
                    </div>
                    <div
                      aria-label="Leave space"
                      class="mx_AccessibleButton mx_IconizedContextMenu_option_red mx_IconizedContextMenu_item"
                      data-test-id="leave-option"
                      role="menuitem"
                      tabindex="-1"
                    >
                      <span
                        class="mx_IconizedContextMenu_icon mx_SpacePanel_iconLeave"
                      />
                      <span
                        class="mx_IconizedContextMenu_label"
                      >
                        Leave space
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        }
      >
        <RovingTabIndexProvider
          handleHomeEnd={true}
          handleUpDown={true}
          onKeyDown={[Function]}
        >
          <div
            className="mx_ContextualMenu_wrapper"
            onClick={[Function]}
            onContextMenu={[Function]}
            onKeyDown={[Function]}
            style={
              Object {
                "bottom": undefined,
                "right": undefined,
              }
            }
          >
            <div
              className="mx_ContextualMenu_background"
              onClick={[Function]}
              onContextMenu={[Function]}
              style={Object {}}
            />
            <div
              className="mx_ContextualMenu"
              hasBackground={true}
              onFinished={[MockFunction]}
              role="menu"
              style={Object {}}
            >
              <div
                className="mx_IconizedContextMenu mx_SpacePanel_contextMenu mx_IconizedContextMenu_compact"
              >
                <div
                  className="mx_SpacePanel_contextMenu_header"
                >
                  test space
                </div>
                <IconizedContextMenuOptionList
                  first={true}
                >
                  <div
                    className="mx_IconizedContextMenu_optionList"
                  >
                    <IconizedContextMenuOption
                      iconClassName="mx_SpacePanel_iconHome"
                      label="Space home"
                      onClick={[Function]}
                    >
                      <MenuItem
                        className="mx_IconizedContextMenu_item"
                        label="Space home"
                        onClick={[Function]}
                      >
                        <RovingAccessibleButton
                          aria-label="Space home"
                          className="mx_IconizedContextMenu_item"
                          onClick={[Function]}
                          role="menuitem"
                        >
                          <AccessibleButton
                            aria-label="Space home"
                            className="mx_IconizedContextMenu_item"
                            element="div"
                            inputRef={
                              Object {
                                "current": <div
                                  aria-label="Space home"
                                  class="mx_AccessibleButton mx_IconizedContextMenu_item focus-visible"
                                  data-focus-visible-added=""
                                  role="menuitem"
                                  tabindex="0"
                                >
                                  <span
                                    class="mx_IconizedContextMenu_icon mx_SpacePanel_iconHome"
                                  />
                                  <span
                                    class="mx_IconizedContextMenu_label"
                                  >
                                    Space home
                                  </span>
                                </div>,
                              }
                            }
                            onClick={[Function]}
                            onFocus={[Function]}
                            role="menuitem"
                            tabIndex={0}
                          >
                            <div
                              aria-label="Space home"
                              className="mx_AccessibleButton mx_IconizedContextMenu_item"
                              onClick={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              role="menuitem"
                              tabIndex={0}
                            >
                              <span
                                className="mx_IconizedContextMenu_icon mx_SpacePanel_iconHome"
                              />
                              <span
                                className="mx_IconizedContextMenu_label"
                              >
                                Space home
                              </span>
                            </div>
                          </AccessibleButton>
                        </RovingAccessibleButton>
                      </MenuItem>
                    </IconizedContextMenuOption>
                    <IconizedContextMenuOption
                      iconClassName="mx_SpacePanel_iconExplore"
                      label="Explore rooms"
                      onClick={[Function]}
                    >
                      <MenuItem
                        className="mx_IconizedContextMenu_item"
                        label="Explore rooms"
                        onClick={[Function]}
                      >
                        <RovingAccessibleButton
                          aria-label="Explore rooms"
                          className="mx_IconizedContextMenu_item"
                          onClick={[Function]}
                          role="menuitem"
                        >
                          <AccessibleButton
                            aria-label="Explore rooms"
                            className="mx_IconizedContextMenu_item"
                            element="div"
                            inputRef={
                              Object {
                                "current": <div
                                  aria-label="Explore rooms"
                                  class="mx_AccessibleButton mx_IconizedContextMenu_item"
                                  role="menuitem"
                                  tabindex="-1"
                                >
                                  <span
                                    class="mx_IconizedContextMenu_icon mx_SpacePanel_iconExplore"
                                  />
                                  <span
                                    class="mx_IconizedContextMenu_label"
                                  >
                                    Explore rooms
                                  </span>
                                </div>,
                              }
                            }
                            onClick={[Function]}
                            onFocus={[Function]}
                            role="menuitem"
                            tabIndex={-1}
                          >
                            <div
                              aria-label="Explore rooms"
                              className="mx_AccessibleButton mx_IconizedContextMenu_item"
                              onClick={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              role="menuitem"
                              tabIndex={-1}
                            >
                              <span
                                className="mx_IconizedContextMenu_icon mx_SpacePanel_iconExplore"
                              />
                              <span
                                className="mx_IconizedContextMenu_label"
                              >
                                Explore rooms
                              </span>
                            </div>
                          </AccessibleButton>
                        </RovingAccessibleButton>
                      </MenuItem>
                    </IconizedContextMenuOption>
                    <IconizedContextMenuOption
                      iconClassName="mx_SpacePanel_iconPreferences"
                      label="Preferences"
                      onClick={[Function]}
                    >
                      <MenuItem
                        className="mx_IconizedContextMenu_item"
                        label="Preferences"
                        onClick={[Function]}
                      >
                        <RovingAccessibleButton
                          aria-label="Preferences"
                          className="mx_IconizedContextMenu_item"
                          onClick={[Function]}
                          role="menuitem"
                        >
                          <AccessibleButton
                            aria-label="Preferences"
                            className="mx_IconizedContextMenu_item"
                            element="div"
                            inputRef={
                              Object {
                                "current": <div
                                  aria-label="Preferences"
                                  class="mx_AccessibleButton mx_IconizedContextMenu_item"
                                  role="menuitem"
                                  tabindex="-1"
                                >
                                  <span
                                    class="mx_IconizedContextMenu_icon mx_SpacePanel_iconPreferences"
                                  />
                                  <span
                                    class="mx_IconizedContextMenu_label"
                                  >
                                    Preferences
                                  </span>
                                </div>,
                              }
                            }
                            onClick={[Function]}
                            onFocus={[Function]}
                            role="menuitem"
                            tabIndex={-1}
                          >
                            <div
                              aria-label="Preferences"
                              className="mx_AccessibleButton mx_IconizedContextMenu_item"
                              onClick={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              role="menuitem"
                              tabIndex={-1}
                            >
                              <span
                                className="mx_IconizedContextMenu_icon mx_SpacePanel_iconPreferences"
                              />
                              <span
                                className="mx_IconizedContextMenu_label"
                              >
                                Preferences
                              </span>
                            </div>
                          </AccessibleButton>
                        </RovingAccessibleButton>
                      </MenuItem>
                    </IconizedContextMenuOption>
                    <IconizedContextMenuOption
                      className="mx_IconizedContextMenu_option_red"
                      data-test-id="leave-option"
                      iconClassName="mx_SpacePanel_iconLeave"
                      label="Leave space"
                      onClick={[Function]}
                    >
                      <MenuItem
                        className="mx_IconizedContextMenu_option_red mx_IconizedContextMenu_item"
                        data-test-id="leave-option"
                        label="Leave space"
                        onClick={[Function]}
                      >
                        <RovingAccessibleButton
                          aria-label="Leave space"
                          className="mx_IconizedContextMenu_option_red mx_IconizedContextMenu_item"
                          data-test-id="leave-option"
                          onClick={[Function]}
                          role="menuitem"
                        >
                          <AccessibleButton
                            aria-label="Leave space"
                            className="mx_IconizedContextMenu_option_red mx_IconizedContextMenu_item"
                            data-test-id="leave-option"
                            element="div"
                            inputRef={
                              Object {
                                "current": <div
                                  aria-label="Leave space"
                                  class="mx_AccessibleButton mx_IconizedContextMenu_option_red mx_IconizedContextMenu_item"
                                  data-test-id="leave-option"
                                  role="menuitem"
                                  tabindex="-1"
                                >
                                  <span
                                    class="mx_IconizedContextMenu_icon mx_SpacePanel_iconLeave"
                                  />
                                  <span
                                    class="mx_IconizedContextMenu_label"
                                  >
                                    Leave space
                                  </span>
                                </div>,
                              }
                            }
                            onClick={[Function]}
                            onFocus={[Function]}
                            role="menuitem"
                            tabIndex={-1}
                          >
                            <div
                              aria-label="Leave space"
                              className="mx_AccessibleButton mx_IconizedContextMenu_option_red mx_IconizedContextMenu_item"
                              data-test-id="leave-option"
                              onClick={[Function]}
                              onFocus={[Function]}
                              onKeyDown={[Function]}
                              onKeyUp={[Function]}
                              role="menuitem"
                              tabIndex={-1}
                            >
                              <span
                                className="mx_IconizedContextMenu_icon mx_SpacePanel_iconLeave"
                              />
                              <span
                                className="mx_IconizedContextMenu_label"
                              >
                                Leave space
                              </span>
                            </div>
                          </AccessibleButton>
                        </RovingAccessibleButton>
                      </MenuItem>
                    </IconizedContextMenuOption>
                  </div>
                </IconizedContextMenuOptionList>
              </div>
            </div>
          </div>
        </RovingTabIndexProvider>
      </Portal>
    </ContextMenu>
  </IconizedContextMenu>
</SpaceContextMenu>
`;
